<!--
  The public master for pairing connections. Externally paired
  devices should connect to this ros master where they can
  find the publicly exposed connections of the robot.
 -->
<launch>
  <!-- ************************* Public Arguments ****************************** -->
  <arg name="auto_invite" default="false"/>  <!-- Take control of the private app manager automatically (should not be true if doing multi-robot) -->
  <arg name="gateway_network_interface" default="$(optenv GATEWAY_NETWORK_INTERFACE)"/>  <!-- If you have multiple network interfaces and want to lock it on one (e.g. 'eth0') -->

  <!-- ********************************* Hub *********************************** -->
  <include file="$(find rocon_hub)/launch/hub.launch">
    <arg name="hub_name" value="Internal Hub" />
    <arg name="hub_port" value="6380" />
    <arg name="zeroconf" value="false" />
  </include>

  <!-- ******************************* Zeroconf ******************************** -->
  <!--  This is for android tablets so robot remocon can autodiscover the public master -->
  <arg name="zeroconf_configuration_file" default="$(find rocon_app_manager)/param/zeroconf.yaml"/>
  <!-- This has to match the name in paired_private for the app manager (redundancy!) -->
  <!-- Other option would be to move the zeroconf node to the private master, but it is not -->
  <!-- necessary there. -->
  <arg name="robot_name"  default="cybernetic_pirate"/>
  <node pkg="zeroconf_avahi" type="zeroconf" name="zeroconf_avahi" ns="zeroconf">
    <rosparam param="services" subst_value="true">
      [ { name: "$(arg robot_name)", type: _ros-master._tcp, port: 11311, domain: local, description: "ros master" } ]
    </rosparam>
  </node>

  <!-- ******************************** Gateway ******************************** -->
  <!--
    The public gateway does not use zeroconf, it needs to know where it can find the internal
    hub shared by private and public ros masters.
   -->
  <node pkg="rocon_gateway" type="gateway.py" name="gateway">
    <rosparam command="load" file="$(find rocon_gateway)/param/default.yaml" />
    <rosparam command="load" file="$(find rocon_gateway)/param/default_blacklist.yaml" />
    <rosparam command="load" file="$(find rocon_app_manager)/param/paired_master_pulls.yaml" />
    <rosparam param="hub_uri">http://localhost:6380</rosparam>
    <rosparam param="name">Public Gateway</rosparam>
    <rosparam param="watch_loop_period">5</rosparam>
    <rosparam param="firewall">false</rosparam>
    <rosparam param="disable_zeroconf">true</rosparam>
    <rosparam param="network_interface" subst_value="true">$(arg gateway_network_interface)</rosparam>
  </node>

  <!--
    Manages the pairing ecosystem, i.e. makes sure the invitation and
    application namespace setting part of the process works smoothly.
  -->
  <node pkg="rocon_app_manager" type="pairing_master.py" name="pairing_master">
    <remap from="pairing_master/gateway_info" to="gateway/gateway_info"/>
    <remap from="pairing_master/remote_gateway_info" to="gateway/remote_gateway_info"/>
    <remap from="pairing_master/set_watcher_period" to="gateway/set_watcher_period"/>
    <param name="auto_invite" value="$(arg auto_invite)"/>
  </node>
</launch>
